rfcs/0032-band-join.md

```diff
@@ -0,0 +1,81 @@
+- If we have an equal condition `A.a = B.b` and its selectivity is high, we can use this condition to distribute the data to acquire parallelism. For `A.p between B.d - 10 and B.d + 20`, we can construct an internal table with order key = `A.a, A.p, A.rid` for A side and order key `B.b, B.d, B.rid` for B side. For `A.q between B.e - 10 and B.e + 20`, we can construct an internal table with order key = `A.a, A.q, A.rid` for A side and order key `B.b, B.e, B.rid` for B side. When a row came from B with (B.b, B.d, B.e) = (100, 200, 300). We can lookup A's internal table row ids with range queries: A between (A.a = 100, A.p = 200 - 10 = 190) and (A.a = 100, A.p = 200 + 20 = 220). Merge the other A's internal table row ids with range queries: A between (A.a = 100, A.q = 300 - 5 = 295) and (A.a = 100, A.p = 200 + 15 = 315). Finally we can intersect the row ids to get the corresponding A matched rows. When a row comes from A, we first need to reverse the range condition as we mentioned before and then do the same logic as for row came from B. Row deleted is basically equivalent to the insertion, but with opposed operators. Update can be handled as delete followed by insert.
+- If we don't have equal conditions `A.a = B.b`, we can only use singleton for both input sides or broadcast one side to the other side. The other logic is basically equivalent with the above example without `A.a` and `B.b` as their prefix keys.
+
+As we can see, we need to estimate the selectivity of the equal condition to decide whether to convert the join into a `BandJoin`, for the reason that we have no statistics now, we can provide a join hint or other ways for users to use it or not. If we have no equal condition, we can always turn the join into `BandJoin`.
+
+### Batch
+
+For batch query, it is more flexible to implement the `BandJoin`. We can even support `RangeJoin` rather than `BandJoin`. One of the most elegant ways to implement it is to use kd-tree as an index, for more details you can read the paper of `A scalable and generic approach to range joins`. At least we have nested-loop-join to run the `BandJoin` or `RangeJoin`, despite their inefficiency.
+
+
+## Unresolved questions
+
+* Parallelize the streaming `BandJoin` if there is no equal condition. Some papers have discussed how to optimize the distributed `BandJoin` by partitioning the input datasets, but it might need a special shuffle algorithm.
+* For simplicity, we can only support `InnerJoin` for now.
+
+## Future possibilities
+
+If you are familiar with Flink, we can find that they have interval join which is just a special case of the `BandJoin`. Interval join requires an equal condition and the range condition looks like `b.timestamp ∈ [a.timestamp + lowerBound; a.timestamp + upperBound]`. Interval join also requires the input stream to be append-only and can cooperate with the watermark to prune the old states. We can use `BandJoin` to implement interval join in the future.
```

Comment:
Consider using the time-bound condition to do state cleaning, there is a unresolve problem: should the time column be placed before join key or after join key?

1. time column before join key: state cleanning is easy (one range delete), but random acesss by join key become impossible.
2. join key before time column: Join key access is fast as usual, but state cleanning's cost is O(N) where N = number of join keys.
3. Alternatively, use two seperated state tables, but this would increase the IO cost multiple times.

Neither sounds a good solution. We need to take a look on Flink's implementation.

Reply:
I had taken a look at Flink's `TimeIntervalJoin` and found that they use alternative two. First they utilize the join equal keys to partition the join state. Second, they maintain a `leftCache` and `rightCache` with type `MapState<Long, List<Tuple2<RowData, Boolean>>>` for each join equal key. The `Long` type is actually the timestamp. BTW, they suffer from a lack of ordered key MapState. The state cleaning mechanism of Flink `TimeIntervalJoin` works like that, it registers a cleanup timer to its time service. The cleanup timer triggered timestamp is calculated based on the row timestamp, input watermark and time interval provided by the user. As soon as the cleanup timer triggers, it iterates the LeftCache and RightCache, deletes expired rows, and registers another cleanup timer based on the earliest valid row timestamp left. As we can see, 
Flink uses a bunch of point deletes triggered by the cleanup timer, rather than a range delete.

Reply:
https://github.com/apache/flink/blob/master/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/interval/TimeIntervalJoin.java

Reply:
> Consider using the time-bound condition to do state cleaning, there is a unresolve problem: should the time column be placed before join key or after join key?
> 
> 1. time column before join key: state cleanning is easy (one range delete), but random acesss by join key become impossible.
> 2. join key before time column: Join key access is fast as usual, but state cleanning's cost is O(N) where N = number of join keys.
> 3. Alternatively, use two seperated state tables, but this would increase the IO cost multiple times.
> 
> Neither sounds a good solution. We need to take a look on Flink's implementation.

I think the first solution is ridiculous. We do not need a fast state cleanning.